### DISCUSSION TITLE
ESP32 vs ESP32S3 Free Heap and RAM

### DISCUSSION THREAD

**Discussion by pavelgolikov** [micropython/micropython#10528] [ESP32] [answered]
Hi everyone.

I have compiled ulab with micropython (https://github.com/v923z/micropython-ulab) according to instructions on that repo to two micro-controllers:

ESP32-S3-WROOM-1 and ESP32-WROOM-32E.

When I was compiling I used GENERIC_S3 for ESP32-S3 and GENERIC for ESP32 as "board" variable.
Then I tested the available heap memory for both with the following script:

```
import time, gc
while True:
    print(gc.mem_free())
    time.sleep(1)
```

Results I get are as follows:

For ESP32-S3-WROOM-1:
164032

for ESP32-WROOM-32E:
106560

What puzzles me is that according to respective documentations ESP32-S3 has 512 KB SRAM (https://www.espressif.com/sites/default/files/documentation/esp32-s3-wroom-1_wroom-1u_datasheet_en.pdf) while ESP32 has 520 KB SRAM (https://www.espressif.com/sites/default/files/documentation/esp32-wroom-32e_esp32-wroom-32ue_datasheet_en.pdf). How come I am getting bigger heap on S3? My expectation was that I will have more memory on ESP32 that on ESP32-S3. Am I misunderstanding something about how memory is allocated with Micropython? Did I do a mistake when I was compiling Micropython for both boards?

Best Regards,

Pavel.

**Comment by jimmo** [comment, marked as answer]
@pavelgolikov The micropython heap is does not take up the full SRAM because we need to leave some behind for the Espressif IDF (e.g. for the WiFi and BLE stacks, as well as things like SSL and socket buffers). There's some history here...

The summary of the heuristic we use is that we ask the IDF for the largest contiguous free region, and then use half of that.

So what you're seeing here is that due to memory usage by the IDF leading to fragmentation, the size of the largest contiguous region is different.

**Reply by pavelgolikov** [reply]
I see. Thank you. So in my case it seems that IDF on ESP32-S3 has had less fragmentation than IDF on ESP-32 and allowed for larger contiguous region? Is that why I am getting more on ESP32-S3 compared to ESP-32?

Do you think uploading different firmware (for example, non-generic esp-32) will allow me to extract more memory compared to what I got now? I feel entitled :) to have more on ESP-32 compared to S3 because of raw RAM size.

**Reply by jimmo** [reply]
> So in my case it seems that IDF on ESP32-S3 has had less fragmentation than IDF on ESP-32 and allowed for larger contiguous region? Is that why I am getting more on ESP32-S3 compared to ESP-32?

Yes that would be my guess.

You can use

```py
>>> import esp32
>>> esp32.idf_heap_info()
...
```

to learn more about the sram usage from the IDF's perspective. (Forgot to mention this in my first reply sorry).

See https://github.com/micropython/micropython/blob/master/ports/esp32/modesp32.c#L181 to understand the output of this.

**Reply by pavelgolikov** [reply]
Thank you jimmo!
